crates/integrations/datafusion/src/physical_plan/repartition.rs

```diff
@@ -0,0 +1,966 @@
+use tracing;
+
+use crate::physical_plan::project::PARTITION_VALUES_COLUMN;
+
+/// Creates an Iceberg-aware repartition execution plan that optimizes data distribution
+/// for parallel processing while respecting Iceberg table partitioning semantics.
+///
+/// Automatically determines the optimal partitioning strategy based on the table's
+/// partition specification.
+///
+/// ## Partitioning Strategies
+///
+/// - **Unpartitioned tables** – Uses round-robin distribution to balance load across workers.
+/// - **Hash partitioning** – Applied for tables with identity or bucket transforms, ensuring
+///   co-location of related data for efficient file clustering. This is used when the `_partition`
+///   column is present AND the partition spec has hash-friendly transforms (Identity/Bucket),
+///   or when source columns with these transforms are available.
+/// - **Round-robin partitioning** – Used for temporal transforms (Year, Month, Day, Hour),
+///   Truncate, or other transforms that don't provide good hash distribution.
+/// - **Mixed transforms** – Combines multiple transform types, using hash partitioning only
+///   when Identity or Bucket transforms are present, otherwise falling back to round-robin.
+///
+/// ## Performance Notes
+///
+/// - Only repartitions when the input scheme or partition count differs from the target.
+/// - Requires an explicit target partition count for deterministic behavior.
+/// - Preserves column order (partitions first, then buckets) for consistent layout.
+///
+/// # Arguments
+///
+/// * `input` – The input execution plan providing data to be repartitioned. For partitioned tables,
+///   the input should include the `_partition` column (added via `project_with_partition`).
+/// * `table_metadata` – Iceberg table metadata containing partition spec.
+/// * `target_partitions` – Target number of partitions for parallel processing (must be > 0).
+///
+/// # Returns
+///
+/// An execution plan that applies the optimal partitioning strategy, or the original input plan if no repartitioning is needed.
+///
+/// # Example
+///
+/// ```ignore
+/// let repartitioned_plan = repartition(
+///     input_plan,
+///     table.metadata_ref(),
+///     4, // Explicit partition count
+/// )?;
+/// ```
+#[allow(dead_code)]
+pub(crate) fn repartition(
+    input: Arc<dyn ExecutionPlan>,
+    table_metadata: TableMetadataRef,
+    target_partitions: NonZeroUsize,
+) -> DFResult<Arc<dyn ExecutionPlan>> {
+    let partitioning_strategy =
+        determine_partitioning_strategy(&input, &table_metadata, target_partitions)?;
+
+    if !needs_repartitioning(&input, &partitioning_strategy) {
+        return Ok(input);
+    }
+
+    Ok(Arc::new(RepartitionExec::try_new(
+        input,
+        partitioning_strategy,
+    )?))
+}
+
+/// Returns whether repartitioning is actually needed by comparing input and desired partitioning
+fn needs_repartitioning(input: &Arc<dyn ExecutionPlan>, desired: &Partitioning) -> bool {
```

Comment:
I'm confused why we need this part. For partitioned table, the `input` is expected to be a `ProjectExec`, and we should always add the partitioning. For unpartitioned table, we should simply return `RoundRobinBatch`